text/0060-enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment/0060-enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment.md

```diff
@@ -0,0 +1,89 @@
+# Enable creating and modifying of FilesContainers within the SAFE Browser environment
+
+- Status: proposed
+- Type: new feature / enhancement
+- Related components: safe-browser, safe-api, safe-nodejs
+- Start Date: 2019-02-11
+- Discussion: https://safenetforum.org/t/enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment/31088
+- Supersedes: (fill me in with a link to RFC this supersedes - if applicable)
+- Superseded by: (fill me in with a link to RFC this is superseded by - if applicable)
+
+## Summary
+
+The NodeJS libraries work via the existing SAFE network CRUD conventions: we take files, we upload them to a network and generate an XORURL we can reference as a result. In the browser we do not have access to a native file sytem, so the underlying NodeJS libraries which the browser calls must allow the uploading of raw bytes, and the creation of empty FilesContainers.
+
+## Conventions
+- The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](http://tools.ietf.org/html/rfc2119).
+
+## Motivation
+
+At the moment, there is no browser native way to create an empty FilesContainer (since the previous convention was the creation of an NFS container which is no longer possible).
+
+## Detailed design
+
+At the moment, the following API calls are available (and implemented in the [NodeJS rust library](https://github.com/maidsafe/safe-nodejs/blob/master/native/src/lib.rs)) with regards to FilesContainers:
+
+* `files_container_create`
+* `files_container_sync`
+* `files_container_get`
+* `files_container_add`
+* `files_container_add_from_raw`
+
+Logic would dictate that the browser would call `files_container_create` to create the files container, however first argument to this method `location: &str` requires a path to a directory or file which will be uploaded to the FilesContainer. The same issues also exist for `files_container_sync` and `files_container_add`.
```

Comment:
Here I would say that we should probably just make that `location` optional. That keeps the `create` function simple, but also more true (otherwise I’d say `create` should always create empty and we should have another func for `from_path` eg. But i think we can keep this API simple and just make this optional 👍)